### ISSUE TITLE
Testing the Ignore method - doesn't work if an exception is thrown synchronously.

### ISSUE THREAD

**Issue by bboyle1234** [dotnet/orleans#4731]
I found a situation where the Ignore() method fails to catch exceptions and demonstrated it with the code below.

Method `Works` is marked with the `async` keyword and the `Ignore` function is executed before its exception is thrown, so this one has a succesfully-ignored exception.

Method `Explodes` does NOT have the `async` keyword, and its exception is thrown before code execution has returned and run the `Ignore` function. So it blows up despite use of `Ignore`. 

In my opinion, this is a major "Gotcha" that will catch developers by surprise. (It caught me) Thoughts? 

```C#
using System;
using System.Diagnostics.CodeAnalysis;
using System.Net;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp3 {
    class Program {

        static void Main(string[] args) {
            MainAsync(args).GetAwaiter().GetResult();
        }

        static Task MainAsync(string[] args) {
            Works().Ignore();
            Explodes().Ignore();
            Console.ReadLine();
            return Task.CompletedTask;
        }

        static async Task Works() {
            throw new Exception("BOOM!");
            await new WebClient().DownloadStringTaskAsync("https://google.com");
        }

        static Task Explodes() {
            throw new Exception("BOOM!");
            return new WebClient().DownloadStringTaskAsync("https://google.com");
        }
    }

    public static class PublicOrleansTaskExtensions {
        private static readonly Action<Task> IgnoreTaskContinuation = t => { _ = t.Exception; };
        [SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "ignored")]
        public static void Ignore(this Task task) {
            if (task.IsCompleted) {
                _ = task.Exception;
            } else {
                task.ContinueWith(
                    IgnoreTaskContinuation,
                    CancellationToken.None,
                    TaskContinuationOptions.NotOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
                    TaskScheduler.Default);
            }
        }
    }
}
```

**Comment by bboyle1234** [comment]
Based on what I have seen here, we have two choices: 

1. Assume that all Orleans contributors understand the deeper implications and weaknesses of the `Ignore` method, and that they will never try to ignore execute a task that throws an exception before yielding execution AND assume that all 3rd party software developers who see and use it as a convenience method in their own application have the same insight, 

OR,

2. Get rid of the `Ignore` method and provide this utility method instead, noting that it is NOT an extension method and cannot be subject to the same gotcha mistake as the `Ignore` method.

public static void FireAndForget(Func<Task> task) {
    Task.Run(async () => { try { await task(); } catch { } });
}

**Comment by bboyle1234** [comment]
Very interesting reply from Stephen Cleary, plus links to some of his articles: 

https://github.com/StephenCleary/AsyncEx/issues/140#issuecomment-401658098

Conclusion ... Orleans will be ok using `Ignore` as-is IF the guidelines presented here have been followed in the Orleans code base. https://blog.stephencleary.com/2016/12/eliding-async-await.html

As for people like me writing Orleans apps - well we are likely to use it wrongly until we learn.

**Comment by jason-bragg** [comment]
@bboyle1234 As Stephen described, this is a subtlety of using tasks which requires developers to understand the behavior for proper use of tasks.  I am aware of no way around this.  At times we've provided 'some' mitigations in our extensibility points (for instance, see WrapException in https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/Lifecycle/LifecycleSubject.cs) but we’ve not done this consistently, and even if we did, it wouldn’t save developers from making similar mistakes in their own code.

The suggested FireAndForget call is not a bad idea, imo, but in general adding custom task management logic is not what Orleans provides.  Capabilities like Ignore() are legacy features we'd prefer not have in Orleans public surface, but are kept around for backwards compatibility reasons.  Such behaviors should be part of the task libraries or some task utility package, not Orleans.
